### PR TITLE
Reinstate https://github.com/inducer/grudge/pull/19

### PR DESCRIPTION
**Pull request by inducer**
cc @nchristensen

### CODE REVIEW

File: grudge/execution.py [hunk 1]
```diff
@@ -525,14 +525,19 @@ def prg(nmatrices):
...
-            matrices_ary_dev = self.array_context.from_numpy(matrices_ary)
+            # Cache operator
+            cache_key = "diff_batch", in_grp, out_grp, tuple(insn.operators),\
+                field.dtype
+            try:
+                matrices_ary_dev = self.bound_op.operator_data_cache[cache_key]
+            except KeyError:
+                matrices = repr_op.matrices(out_grp, in_grp)
+                matrices_ary = np.empty((
+                    noperators, out_grp.nunit_dofs, in_grp.nunit_dofs))
```
Comment (inducer) [inline grudge/execution.py:536]:
```suggestion
                    noperators, out_grp.nunit_dofs, in_grp.nunit_dofs), dtype=field.entry_dtype)
```

**Comment by nchristensen** [comment]
How can I replicate this locally? I just ran the tests on Porter without any problem.

**Comment by inducer** [comment]
As I said in #19, running `ulimit -v 3000000` beforehand allowed me to trigger it reliably.

**Comment by nchristensen** [comment]
It seems to be system dependent. On Monza failures start arount 1800000 and on Porter they start around 2200000 for me.

**Comment by inducer** [comment]
Had a theory: Without the periodic (synchronous!) memory transfers (that are eliminated by this PR), the CL enqueue front can (and, for POCL, will!) essentially run arbitrarily far ahead, consuming an arbitrarily large amount of memory that gets preallocated for output buffers. Periodic synchronization should fix this... the question is: who should be responsible for that? Maybe POCL? FWIW, this also explains why CL logging helps: it slows down the enqueue front to the point where execution can keep up.

cc @fischer1 @matthiasdiener

**Comment by inducer** [comment]
Adding synchronization helps:
```diff
--- a/test/test_grudge.py
+++ b/test/test_grudge.py
@@ -801,6 +801,12 @@ def test_convergence_advec(actx_factory, mesh_name, mesh_pars, op_type, flux_typ
 
 # {{{ models: maxwell
 
+def acf():
+    context = cl._csc()
+    queue = cl.CommandQueue(context)
+    return PyOpenCLArrayContext(queue)
+
+
 @pytest.mark.parametrize("order", [3, 4, 5])
 def test_convergence_maxwell(actx_factory,  order):
     """Test whether 3D Maxwell's actually converges"""
@@ -851,6 +857,7 @@ def test_convergence_maxwell(actx_factory,  order):
 
         step = 0
         for event in dt_stepper.run(t_end=final_t):
+            actx.queue.finish()
             if isinstance(event, dt_stepper.StateComputed):
                 assert event.component_id == "w"
                 esc = event.state_component
```
That still leaves us with the problem of finding a central place where we can judiciously add some synchronization. It seems less plausible to offload this responsibility on the user, given how long it's taken us to figure out what's going on.

**Comment by inducer** [comment]
Hrmph, https://github.com/inducer/meshmode/pull/60 didn't suffice. Next attempt: https://github.com/inducer/meshmode/pull/61

**Comment by inducer** [comment]
#61 seems to have helped. In it goes.

:tada:

**Pull request merged**